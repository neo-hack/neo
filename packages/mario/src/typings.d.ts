declare module 'gulp-copy'
declare module 'js-yaml-js-types/all'
declare module 'js-yaml-js-types'
