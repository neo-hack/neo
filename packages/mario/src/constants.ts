export const LIFE_CYCLES = {
  START: 'start',
  STEP: 'step',
}
