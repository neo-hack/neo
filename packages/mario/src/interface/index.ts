import type Debug from 'debug'

export type WebhookEvent = string

export interface Step {
  /**
   * A name for your step to display on GitHub.
   */
  name?: string
  /**
   * TODO: dynamic load action not support now
   * Selects an action to run as part of a step in your job. An action is a reusable unit of code. You can use follow built-in actions
   * - replace
   * - copy
   * - mv
   * - rm
   */
  uses?: string
  /**
   * A map of the input parameters defined by the action. Each input parameter is a key/value pair. Input parameters are set as environment variables. The variable is prefixed with INPUT_ and converted to upper case.
   */
  with?: {
    [k: string]: unknown
  }
  /**
   * TODO: env setup not support now
   * Sets environment variables for steps to use in the virtual environment. Public actions may specify expected environment variables in the README file. If you are setting a secret in an environment variable, you must set secrets using the secrets context.
   */
  env?: {
    [k: string]: unknown
  }
  /**
   * TODO: condition not support
   * Identifies any steps that must complete successfully before this step will run. It can be a string or an array of strings. If a step fails, all steps that need it will also fail unless the steps use a conditional statement that causes the step to continue.
   */
  if?: string
  /**
   * Runs command line programs using the operating system's shell. If you do not provide a name, the step name will default to the run command. Commands run using non-login shells by default.
   */
  run?: string
  /**
   * TODO: working-directory not support
   * The default directory that the action uses in a job's workspace.
   */
  'working-directory'?: string
  /**
   * Prevents a job from failing when a step fails. Set to true to allow a job to pass when this step fails.
   */
  'continue-on-error'?: boolean
}

/**
 * Each job must have an id to associate with the job. The key job_id is a string and its value is a map of the job's configuration data. You must replace <job_id> with a string that is unique to the jobs object. The <job_id> must start with a letter or _ and contain only alphanumeric characters, -, or _.
 */
export interface Job {
  /**
   * The name of the job displayed on GitHub.
   */
  name?: string
  /**
   * Path patterns of the job
   */
  paths?: string
  /**
   * Identifies any jobs that must complete successfully before this job will run. It can be a string or array of strings. If a job fails, all jobs that need it will also fail unless the jobs use a conditional statement that causes the job to continue.
   */
  /**
   * A job contains a sequence of tasks called steps. Steps can run commands, run setup tasks, or run an action in your repository, a public repository, or an action published in a Docker registry. Not all steps run actions, but all actions are run as a step. Each step runs in its own process in the virtual environment and has access to the workspace and filesystem. Because steps are run in their own process, changes to environment variables are not preserved between steps. GitHub provides built-in steps to set up and complete a job.
   */
  steps?: Step[]
}

export interface Workflow {
  /**
   * The name of your workflow. GitHub displays the names of your workflows on your repository's actions page. If you omit this field, GitHub sets the name to the workflow's filename.
   */
  name?: string
  /**
   * A workflow run is made up of one or more jobs. Jobs run in parallel by default. To run jobs sequentially, you can define dependencies on other jobs using the jobs.<job_id>.needs keyword.
   */
  jobs?: Record<string, Job>
}

export interface Context {
  cwd: string
  debug: Debug.Debugger
  variables?: Record<string, any>
}

export type Action<T = Record<string, string>> = (args: T, ctx: Context) => any
