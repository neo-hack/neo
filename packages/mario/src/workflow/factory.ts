import {
  isObject,
  isString,
  mapValues,
  template,
  templateSettings,
} from 'lodash-es'

import type { Action, Context } from '../interface'

templateSettings.interpolate = /\${{([\s\S]+?)}}/g

/**
 * @description tpl format like `${{ inputs.variable }}`
 */
export const tpl = (templateString: string, variable?: Record<string, any>) => {
  if (!variable || !isString(templateString)) {
    return templateString
  }
  const compiled = template(templateString)
  return compiled(variable)
}

export const mapValuesDeep = (
  v: Record<string, any>,
  callback: (value: string) => any,
): Record<string, any> =>
  Array.isArray(v)
    ? v.map(v => mapValuesDeep(v, callback))
    : isObject(v)
      ? mapValues(v, v => mapValuesDeep(v, callback))
      : callback(v)

export const wrapOptions = (options: Record<string, any>, variables: Context['variables']) => {
  return mapValuesDeep(options, v => tpl(v, variables))
}

export const factory = (action: Action<any>): Action<any> => {
  return (options, ctx) => {
    const next = wrapOptions(options, ctx.variables)
    return action(next, ctx)
  }
}
