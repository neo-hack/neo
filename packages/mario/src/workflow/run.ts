import Shell from 'gulp-shell'

import type { Action } from '../interface'

export interface RunOptions {
  commands?: string[] | string
  quiet?: boolean
  ignoreErrors?: boolean
}

export const run: Action<RunOptions> = ({ commands = [], ...options }, ctx) => {
  return Shell(commands, { cwd: ctx.cwd, ...options })
}
