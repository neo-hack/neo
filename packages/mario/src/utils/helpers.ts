import { Workflow } from '../interface'
import { hooks } from './hooks'

import Observable from 'zen-observable'
import type Listr from 'listr'
import { LIFE_CYCLES } from 'src/constants'

/**
 * @description convert schema to listr
 */
export const toListr = (schema: Workflow) => {
  const tasks: Listr.ListrTask[] = Object.keys(schema.jobs || {}).map((key) => {
    const taskName = schema.jobs![key].name || key
    return {
      title: taskName,
      task: () => {
        let ob: ZenObservable.SubscriptionObserver<any>
        const OB = new Observable<any>((observer) => {
          ob = observer
        })
        hooks.addHooks({
          [LIFE_CYCLES.STEP]: (args: { step: string }) => {
            ob.next(args.step)
          },
          [taskName]: (args: { error: Error }) => {
            if (!args.error) {
              ob.complete()
              return
            }
            ob.error(args.error)
          },
        })
        return OB as any
      },
    }
  })
  return tasks
}
