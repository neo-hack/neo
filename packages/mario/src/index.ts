import { createWorkflow, readWorkflowSchema, CreateWorkflowOptions } from './workflow'
import { pick } from 'lodash-es'
import { Workflow } from './interface'

export { hooks } from './utils/hooks'
// export { LogLevel } from 'consola'

export const create = async (
  filepath: string,
  options: Omit<CreateWorkflowOptions, 'schema'> & {
    jobs?: string[]
  } = {},
) => {
  const schema = await readWorkflowSchema(filepath)
  if (options.jobs) {
    schema.jobs = pick(schema.jobs, options.jobs) as Workflow['jobs']
  }
  const workflow = await createWorkflow({ schema, ...options })
  return {
    schema,
    start: async () => {
      await workflow()
    },
  }
}
