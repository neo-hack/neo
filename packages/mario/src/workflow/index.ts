import readYaml from 'read-yaml-file'
import gulp from 'gulp'
import consola, { Consola } from 'consola'
import filter from 'gulp-filter'
import gulpDebug from 'gulp-debug'

import { hooks } from '../utils/hooks'
import { Workflow, Job, Context, Step } from '../interface'
import { LIFE_CYCLES } from '../constants'
import { builtInUses } from './uses'
import { run, RunOptions } from './run'
import { debug, prefix } from '../utils/logger'

export const readWorkflowSchema = async (filepath: string) => {
  const workflow = await readYaml<Workflow>(filepath)
  return workflow
}

type CreateJobOptions = {
  runAction: (id: string, args: any, options: Step, ctx: Context) => any
  runShell: (args: RunOptions, options: Step, ctx: Context) => any
  // job name fallback
  key: string
  job: Job
  cwd?: string
}

export const createJob = async ({ job, key, ...options }: CreateJobOptions) => {
  return async () => {
    const taskName = job.name || key
    const task = async () => {
      const pipes = async () => {
        let stream = gulp.src(job.paths ? job.paths : ['*'], {
          cwd: options.cwd!,
          allowEmpty: true,
          dot: true,
        })
        stream = stream.pipe(filter(['**', '.**', '!**/node_modules/**']))
        if (process.env.DEBUG?.includes(`${prefix}:job`)) {
          stream = stream.pipe(gulpDebug({ title: taskName }))
        }
        for (const step of job.steps || []) {
          const extra: Step = { 'continue-on-error': step['continue-on-error'] }
          if (!step.uses && !step.run) {
            continue
          }
          if (step.uses) {
            const cb = await options
              .runAction?.(step.uses, step.with, extra, {
                cwd: options.cwd!,
                debug: debug.uses,
              })
              .catch((error: Error) => {
                if (!extra['continue-on-error']) {
                  hooks.callHook(taskName, { job: taskName, error })
                  stream.emit('error')
                } else {
                  stream.emit('success')
                }
              })
            if (!cb) {
              continue
            }
            const cbs = Array.isArray(cb) ? cb : [cb]
            cbs.forEach((cb) => {
              stream = stream.pipe(cb).on('error', function (this: any, error: Error) {
                if (!extra['continue-on-error']) {
                  hooks.callHook(taskName, { job: taskName, error })
                  stream.emit('error')
                } else {
                  stream.emit('success')
                }
              })
            })
          }
          // exec shell
          if (step.run) {
            const cb = options.runShell?.({ commands: step.run, ...step.with }, extra, {
              cwd: options.cwd!,
              debug: debug.run,
            })
            stream = stream.pipe(cb)
          }
          hooks.callHook(LIFE_CYCLES.STEP, { step: step.name })
        }
        stream = stream.pipe(
          gulp.dest((file) => {
            return file.base
          }),
        )
        return stream
      }
      return new Promise((resolve, reject) => {
        pipes().then((stream) => {
          stream.on('end', resolve)
          stream.on('success', resolve)
          stream.on('error', reject)
        })
      }).then(() => {
        hooks.callHook(taskName, { job: taskName })
      })
    }
    gulp.task(taskName || key, task)
    return {
      job: taskName,
    }
  }
}

const runAction: CreateJobOptions['runAction'] = async (id, args, _options, ctx) => {
  debug.uses('run uses %s with %O', id, args)
  const action = builtInUses[id]
  if (!action) {
    consola.error(`${action} not found, dynamic import not support now!`)
    return false
  }
  const cb = await action(args, ctx)
  return cb
}

const runShell: CreateJobOptions['runShell'] = (args, options, ctx) => {
  debug.run('exec command %O', args)
  return run({ ...args, ignoreErrors: options['continue-on-error'] }, ctx)
}

export type CreateWorkflowOptions = {
  schema: Workflow
  cwd?: CreateJobOptions['cwd']
  logLevel?: Consola['level']
  // filter workflow.jobs by job keys
  jobs?: string[]
}

export const createWorkflow = async ({
  schema,
  cwd = process.cwd(),
  ...options
}: CreateWorkflowOptions) => {
  const keys = Object.keys(schema.jobs || {})
  consola.level = options.logLevel ?? consola.level
  return async () => {
    const jobs: string[] = []
    for (const key of keys) {
      if (options.jobs && !options.jobs?.includes(key)) {
        continue
      }
      debug.job('create job %s on cwd %s', key, cwd)
      // TODO: curry is wild
      const job = await createJob({
        job: schema.jobs![key],
        runAction,
        runShell,
        key,
        cwd,
        ...options,
      })
      const { job: name } = await job()
      jobs.push(name)
    }
    gulp.series(...jobs)((error) => {
      consola.error(error)
    })
  }
}
