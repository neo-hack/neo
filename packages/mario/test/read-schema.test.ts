import { readWorkflowSchema } from '../src/workflow'
import { r } from './helpers'

describe('read schema', () => {
  it('read schema takes array as input', async () => {
    const schema = await readWorkflowSchema(r('test/assets/array-input.yaml'))
    expect(schema).toMatchSnapshot()
  })

  it('read complete workflow schema', async () => {
    const schema = await readWorkflowSchema(r('test/assets/workflow.yaml'))
    expect(schema).toMatchSnapshot()
  })

  it('read workflow contain regexp', async () => {
    const schema = await readWorkflowSchema(r('test/assets/regexp.yaml'))
    expect(schema).toMatchSnapshot()
  })
})
