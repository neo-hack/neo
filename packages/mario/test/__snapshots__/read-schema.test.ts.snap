// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read schema read complete workflow schema 1`] = `
Object {
  "jobs": Object {
    "REPLACE": Object {
      "paths": "test/**/*.js",
      "steps": Array [
        Object {
          "name": "REPLACE JS Files",
        },
        Object {
          "action": "replace",
          "with": Object {
            "patterns": Array [
              Object {
                "match": "one",
                "replacement": "one target",
              },
              Object {
                "match": "two",
                "replacement": "two targets",
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`read schema read schema takes array as input 1`] = `
Object {
  "with": Array [
    Object {
      "find": "one",
      "replace": "one target",
    },
    Object {
      "find": "two",
      "replace": "two targets",
    },
  ],
}
`;

exports[`read schema read workflow contain regexp 1`] = `
Object {
  "jobs": Object {
    "REPLACE": Object {
      "paths": "output/**/*.ts",
      "steps": Array [
        Object {
          "name": "Replace Action",
        },
        Object {
          "uses": "replace",
          "with": Object {
            "match": "one",
            "replacement": "target",
          },
        },
        Object {
          "name": "Replace Action",
        },
        Object {
          "uses": "replace",
          "with": Object {
            "pairs": Array [
              Object {
                "match": "two",
                "replacement": "target",
              },
              Object {
                "match": "three",
                "replacement": "target",
              },
            ],
          },
        },
        Object {
          "name": "Replace Action",
        },
        Object {
          "uses": "replace",
          "with": Object {
            "match": /@aiou\\\\/\\*-template/g,
            "replacement": "target",
          },
        },
      ],
    },
  },
}
`;
